drivers/soundwire/intel.c

```diff
@@ -1387,6 +1387,8 @@ static int intel_master_startup(struct sdw_master_device *md)
 
 		pm_runtime_set_active(&md->dev);
 		pm_runtime_enable(&md->dev);
+	} else {
+		dev_pm_set_driver_flags(&md->dev, DPM_FLAG_NEVER_SKIP);
```

Comment:
sorry @bardliao I would need more explanations here. I did not understand the reports in issue #1595 and I am concerned that there are very few users of this flag (see below), and if it's needed here we would also need it for the PCI part when we disable PM runtime.

That said, I am not sure what happens if we disable pm_runtime but there are pretend RT700 devices exposed in the DSDT tables, that are not physically present on the board and are never attached. It could very well be a corner case where this is required, but that would need to be made explicit in the commit message.

@ranj063 @kv2019i @lyakh FYI, if you are familiar with that flag please help on this one.

````
git grep DPM_FLAG_NEVER_SKIP
Documentation/driver-api/pm/devices.rst:        ``DPM_FLAG_NEVER_SKIP`` and ``DPM_FLAG_SMART_PREPARE`` driver power
Documentation/power/pci.rst:The DPM_FLAG_NEVER_SKIP flag prevents the PM core from using the direct-complete
drivers/base/power/main.c:              !dev_pm_test_driver_flags(dev, DPM_FLAG_NEVER_SKIP);
drivers/gpu/drm/amd/amdgpu/amdgpu_kms.c:                dev_pm_set_driver_flags(dev->dev, DPM_FLAG_NEVER_SKIP);
drivers/gpu/drm/i915/intel_runtime_pm.c:        dev_pm_set_driver_flags(kdev, DPM_FLAG_NEVER_SKIP);
drivers/gpu/drm/radeon/radeon_kms.c:            dev_pm_set_driver_flags(dev->dev, DPM_FLAG_NEVER_SKIP);
drivers/misc/mei/pci-me.c:      dev_pm_set_driver_flags(&pdev->dev, DPM_FLAG_NEVER_SKIP);
drivers/misc/mei/pci-txe.c:     dev_pm_set_driver_flags(&pdev->dev, DPM_FLAG_NEVER_SKIP);
drivers/net/ethernet/intel/e1000e/netdev.c:     dev_pm_set_driver_flags(&pdev->dev, DPM_FLAG_NEVER_SKIP);
drivers/net/ethernet/intel/igb/igb_main.c:      dev_pm_set_driver_flags(&pdev->dev, DPM_FLAG_NEVER_SKIP);
drivers/pci/pcie/portdrv_pci.c: dev_pm_set_driver_flags(&dev->dev, DPM_FLAG_NEVER_SKIP |
include/linux/pm.h:#define DPM_FLAG_NEVER_SKIP          BIT(0)
````

Reply:
@plbossart @bardliao the NEVER_SKIP flag comes into picture if we're using direct complete for system suspend aka SMART_SUSPEND. As far as I know, this feature never worked for us (we do not enable it today). So we always resume the SOF audio device before system suspend. Additionally, the placement of this flag setting is a bit questionable since you set NEVER_SKIP only when runtime PM is disabled

Reply:
The purpose of this PR is to make sure intel_suspend() will be invoked in suspend if the link is not runtime suspended. Looking at [power/main.c](https://github.com/thesofproject/linux/blob/integration/soundwire-latest/drivers/base/power/main.c). It will check [if (dev->power.direct_complete)](https://github.com/thesofproject/linux/blob/integration/soundwire-latest/drivers/base/power/main.c#L1720) to decide whether device's suspend callback function will be invoked or not. The direct_complete flag is basically assigned at [device_prepare()](https://github.com/thesofproject/linux/blob/integration/soundwire-latest/drivers/base/power/main.c#L1939). 
At this point, all sdw-master's direct_complete flag are true for the no_pm_callbacks flag is true. And the direct_complete flag of those codecs which are attached are also true for they are runtime suspended. However, for those codecs which are not attached, their direct_complete flag are false for the no_pm_callbacks flag is false and they are not runtime suspended.
It means the unattached codecs' suspend callback function will be invoked and [dpm_clear_superiors_direct_complete()](https://github.com/thesofproject/linux/blob/integration/soundwire-latest/drivers/base/power/main.c#L1784) will be called after that.
At this point, sdw-master's no_pm_callbacks flag will be set to [false](https://github.com/thesofproject/linux/blob/integration/soundwire-latest/drivers/base/power/main.c#L1660)
That's the reason why intel_suspend() will be triggered if there is a unattached codec on the sdw link, but not if there is no unattached codec on the sdw link.
To ensure intel_suspend() will be triggered whenever it is not runtime suspended, we need to set direct_complete flag false. I think the easiest way is to set DPM_FLAG_NEVER_SKIP, but I will check how can we let dev->power.no_pm_callbacks = false.
```
	dev->power.direct_complete = state.event == PM_EVENT_SUSPEND &&
		((pm_runtime_suspended(dev) && ret > 0) ||
		 dev->power.no_pm_callbacks) &&
		!dev_pm_test_driver_flags(dev, DPM_FLAG_NEVER_SKIP);
```

Reply:
@bardliao in the issue #1595, the PCI device and the SoundWire master don't enable PM runtime. The Rt700 is not present so pm_runtime is not enabled either for the codec device.

can you clarify this part:
````
It means the unattached codecs' suspend callback function will be invoked and dpm_clear_superiors_direct_complete() will be called after that.
At this point, sdw-master's no_pm_callbacks flag will be set to false

That's the reason why intel_suspend() will be triggered if there is a unattached codec on the sdw link, but not if there is no unattached codec on the sdw link.
````

I have no idea what the no_pm_callbacks means, really difficult for me to follow the logic.

Reply:
> @bardliao in the issue #1595, the PCI device and the SoundWire master don't enable PM runtime. The Rt700 is not present so pm_runtime is not enabled either for the codec device.

Right, but tr700's system sleep pm is enabled.

> 
> can you clarify this part:
> 
> ```
> It means the unattached codecs' suspend callback function will be invoked and dpm_clear_superiors_direct_complete() will be called after that.
> At this point, sdw-master's no_pm_callbacks flag will be set to false
> 
> That's the reason why intel_suspend() will be triggered if there is a unattached codec on the sdw link, but not if there is no unattached codec on the sdw link.
> ```

[__device_suspend()](https://github.com/thesofproject/linux/blob/integration/soundwire-latest/drivers/base/power/main.c#L1681) will test [dev->power.direct_complete](https://github.com/thesofproject/linux/blob/integration/soundwire-latest/drivers/base/power/main.c#L1720) to decide whether device's pm callback function need to run or not. rt700's power.direct_complete will be false since its pm runtime is disabled.
It means tr700's system sleep suspend callback function will be invoked and more important [dpm_clear_superiors_direct_complete(dev)](https://github.com/thesofproject/linux/blob/integration/soundwire-latest/drivers/base/power/main.c#L1653) will be called and it will clear rt700's parent which is sdw-master's power.direct_complete flag. That's why sdw-master's pm callback function which is intel_suspend() will be triggered if rt700's driver is loaded but pm runtime is disabled.

> 
> I have no idea what the no_pm_callbacks means, really difficult for me to follow the logic.

To my understanding, it means that there is no pm callback functions if no_pm_callback is true.
So the logic is we can directly complete the suspend process if there is no pm callback function for a device.
```
void device_pm_check_callbacks(struct device *dev)
{
	spin_lock_irq(&dev->power.lock);
	dev->power.no_pm_callbacks =
		(!dev->bus || (pm_ops_is_empty(dev->bus->pm) &&
		 !dev->bus->suspend && !dev->bus->resume)) &&
		(!dev->class || pm_ops_is_empty(dev->class->pm)) &&
		(!dev->type || pm_ops_is_empty(dev->type->pm)) &&
		(!dev->pm_domain || pm_ops_is_empty(&dev->pm_domain->ops)) &&
		(!dev->driver || (pm_ops_is_empty(dev->driver->pm) &&
		 !dev->driver->suspend && !dev->driver->resume));
	spin_unlock_irq(&dev->power.lock);
}
```
Ideally, the no_pm_callback flag of sdw-master device should be false, but it is true so far.
[device_pm_check_callbacks()](https://github.com/thesofproject/linux/blob/integration/soundwire-latest/drivers/base/power/main.c#L2086) will be triggered by [device_register()](https://github.com/thesofproject/linux/blob/integration/soundwire-latest/drivers/soundwire/master.c#L50). However, dev->driver is null at that moment, and device_pm_check_callbacks() will not be triggered again after that. So, the no_pm_callback flag of sdw-master device is false.